### DISCUSSION TITLE
`contextvars` requirement

### DISCUSSION THREAD

**Discussion by yangskyboxlabs** [saltstack/salt#66664] [Q&A] [answered]
I'd like to understand the context which leads `salt` to continue to carry a dependency on `contextvars` from PyPI.

`base.txt` simply notes that `salt-ssh` requires it, but I see no unusual usage of `contextvars` in the `salt-ssh` stack contained within Salt.

**Comment by lkubb** [comment, marked as answer]
Salt-SSH works by packing Salt and dependencies into a so-called thin tarball. Essentially, it copies Salt + its hard dependencies from the master into a tar archive that is sent to the target. In case Salt-SSH connects to a target with a Python version of 3.6 or lower, it needs a `contextvars` module that can be packed into the thin tar.

To discover the sources, `salt.utils.thin` imports the module and references `<mod>.__file__`. The stdlib `contextvars` imports from `_contextvars` itself, which does not work with this mechanism, while the library is self-contained.

**Comment by yangskyboxlabs** [comment]
Thanks! That was super useful. This gives me a couple viable routes to explore with our current plan to build some tooling that requires tight integration with a masterless salt minion.
